### ISSUE TITLE
Move block + MultiBody model can give issues with index reduction

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#1523]
**Reported by otter on 21 Jun 2014 13:48 UTC**
The attached model (from Leo Gall) fails to translate in Dymola (and most likely in all other Modelica tools).

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1523

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 21 Jun 2014 14:11 UTC**
Fixed in 27e17cf0c38abd227b405a3fc8fe0f3c32f07196 (trunk) and 9b5c652893da661c70322633e33b18bd3efaf180 (maintenance) by
- Improving block Modelica.Mechanics.Rotational.Sources.Move
- Improving block Modelica.Mechanics.Translational.Sources.Move
- Adding test model ModelicaTest.MultiBody.PlanarLoopWithMove.
The test model is a combination of a multi-body system with a kinematic loop and a Rotational.Sources.Move block, where symbolic transformation fails in MSL 3.2.1 build2:

Standard symbolic transformation works in the following way if functions are present with annotation "InineAfterIndexReduction = true". 

1. These functions are not inlined, the Pantelides algorithm is applied and the equations are aymbolically differentiated. 
2. The functions are inlined and further symbolic transoformation is performed (sorting the equations, dummy derivative method etc.. 

The test model contains the following functions that have annotation "InineAfterIndexReduction = true": 
1. MultiBody.Frames.resolve1/resolve2(..) functions are used to transform the coordinates of a vector in another coordinate system. When these functions need to be differentiated in the Pantelides algorithm, then Eulers differentiation rule is applied with the angular velocity that is usually a factor of 2-3 more efficient as if direct differentation is applied. The drawback is that the sparsity structure is more "coarse" and this lets the Pantelides algorithm fail in this example (this happens usually only in systems with kinematic loops). This means that the differentiated equation system is structurally singular, after the functions with annotation "InineAfterIndexReduction = true" are inlined. 
2. The Rotational.Sources.Move block is using internally functions to express that an input u[2] is the derivative of of input u[1]. In order that this is possible, the functions have the annotation "InineAfterIndexReduction = true" so that the differentiation takes place with the not inlined functions. After the Pantelides algorithm and the differentation is applied, the functions are inlined to enhance efficiency. 
The problem with (1) is that after inlining the functions with annotation "InineAfterIndexReduction = true", the equations are structurally singular, but the source of the singularity is now known. One remedy is to restart symbolic processing and to inline all functions with "InineAfterIndexReduction = true" before the Pantelides algorithm is applied. In the model, this approach is in principal succeessful and the model would be structurally regular. However, the Move block is no longer working correctly if the Pantelides algorithm is appliwed with inlining. For this reason in MSL 3.2.1 build 3, the annotation "InineAfterIndexReduction = true" in the Move block has been replaced by annotation "Inline = false". This results in slightly less efficient code, but is more robust in the situation at hand.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Jan 2016 10:19 UTC**
Re-opened, since this change has the effect that Modelica.Mechanics.Rotational.Examples.GenerationOfFMUs does not longer translate in Dymola.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Jan 2016 13:27 UTC**
Based on an email discussion with Sven Erik Mattsson my current understanding is the following:

The "InlineAfterIndexReduction=true" annotation is or was present in

- Modelica.Mechanics.Rotational.Sources.Move in order handle input signals where the variable u, its first, and its second derivative are provided from the "outside" of Modelica and in Modelica the translator needs to be informed that the derivatives of u are also external signals.
- Modelica.Mechanics.MultiBody.Frames.resolve1/.resolve2 in order to perform differentiation of kinematic equations, without differentiating the orientation matrix (but using Euler's differentiation rule with the angular velocity instead).

The purpose of  "InlineAfterIndexReduction=true" is to perform Pantelides and BLT with the non-inlined form, then inline and use the inlined form for state selection and for better efficiency. Unfortunately, there are a few multi-body systems (usually with kinematic loops), where the structural analysis with the non-inlined form indicates a singular system, although it is not. Once Dymola recognizes this, it tries to inline all functions with "inlineAfterIndexReduction=true" and then applies Pantelides/BLT again (for multi-body systems this means that the efficiency is usually greatly reduced, because the orientation matrix is then differentiated twice; however having an inefficient solution is better as having no solution).

The Move block is defined in such a way that output signal phi = position(u,time) and u is a vector of three input signals(say u[3] = {q,qd=der(q), qdd=der(der(q)) and when phi needs to be differentiated then the provided derivative function of position is utilized. Therefore, to get this differentiation of functions correct, function position must not be inlined, but the provided function position_der and position_der2 used during Pantelides/BLT. For this reason, originally the annotation "InlineAfterIndexReduction=true" was used here.

Now, the problem is that if "inlineAfterIndexReduction" functions are inlined (because "resolved" functions failed), then also the Move functions are inlined too. The result is that a combination of the corresponding multi-body system together with a Move block will fail (for example: ModelicaTest.MultiBody.PlanarLoopWithMove), because the Move functions will fail to work correctly when they are inlined before Pantelides/BLT is applied. Due to this reason the "InlineAfterIndexReduction=true" annotation of the Move block has been replaced by "Inline=false", that is the block is never inlined.

Unfortunately, at that time it was not detected that the Move block will fail if it is not inlined before the state selection algorithm (for the state selection it is important to have the "correct" functional dependency of phi=f(q) and if the Move functions are not inlined then a very bad dependency will make the state selection fail). Due to "Inline=false", the Move block will not be inlined and therefore it will fail in cases where the Move block is involved in state selection, as in model Modelica.Mechanics.Rotational.Examples.GenerationOfFMUs).

The question is what to do? One possibility would be to introduce an additional annotation, say "strictInlineAfterIndexReduction = true" that defines that the corresponding function is NEVER inlined before Pantelides/BLT. This annotation could be defined in the Move block and then the problems above would be solved. However, this can only be done in a new Modelica release and then MSL must be allowed to use this annotation. So this solution would only be applicable in several years and not for MSL 3.2.2. Another (better) possibility would be to introduce a special language element that explicitly defines the relationships of three input signals (that one is the derivative of the other one). Again, such a solution would only be available in several years.

The Move block has the additional feature that the "noDerivative" annotation is used (because the "u" vector argument shall not be differentiated when function position(u) is differentiated). Sven Erik and Hans propose to utilize this feature: If annotation Inline=false is present, then the tool should avoid to inline according to the specification. However, if state selection fails and the function has the "noDerivative" annotation, then this is an indication that a situation as for the Move block is present. It is then proposed to make another try and inline the function before the state selection. However, this approach will not work for the current Move block, because the block has an additional, second dummy argument, "called dummy"). When the Move block is called, actually "time" is given for this dummy argument ("phi = position(u,time)"). I do no longer understand why this dummy argument is needed. Sven Erik and Hans propose to just remove it. If it is removed, then there remains just one argument with the "noDerivative" annotation and the sketched approach can be used.

To summarize, it is proposed to remove the "dummy" argument from the protected position/position_der/position_der2 functions in the Move block. This is a backwards compatible change (because these functions are defined in the protected section). Tools, like Dymola, would have then to implement the strategy sketched above.

If there is no counter proposal or a reason why this solution will lead to yet other issues, I will fix this as indicated.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Jan 2016 13:37 UTC**
Fixed in 807dfa2ebd10e728f97e712fe28f48b90eb9e970 as suggested by removing the "dummy" argument from the internal functions in the protected sections of blocks:
- Modelica.Mechanics.Rotational.Sources.Move
- Modelica.Mechanics.Translational.Sources.Move

This ticket is not closed until tool support (especially Dymola) shows that this fix works as expected.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 19 Jan 2016 08:46 UTC**
I realized that the code-generation will also be influenced in a bad way by this (the trivial functions should not be inlined).

One solution would be to use "LateInline=true" (instead of "Inline=false" or previous "InlineAfterIndexReduction=true") - similarly as for property caching.

However, it still needs to fixed to work as expected and I have not investigated if this may cause other problems.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 21 Jan 2016 07:31 UTC**
Fixed in 3432a2218e1a4ff83dc1bd67f799aac6f6644eb4, according to the suggestion from Hans and Sven Erik:
Updated Move blocks of the Translational and Rotational library:
- Re-introduced the dummy argument (still not yet known why it is here, but just in case, and it does not hurt)
- changed annotation to LateInline=true.
Tested these changes with all test models in Modelica and ModelicaTest where the Move blocks are used (with Dymola 2017 dev4)

**Comment by modelica-trac-importer** [comment]
**Changelog modified by otter on 21 Jan 2016 07:31 UTC**
Usage of Move blocks of Rotational or Translational library sometimes failed. This has been fixed

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 21 Jan 2016 09:29 UTC**
Are there examples in Modelica(Test?) for testing this or do we need to use the attached models?

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 21 Jan 2016 09:36 UTC**
Replying to [comment:7 sjoelund.se]:
> Are there examples in Modelica(Test?) for testing this or do we need to use the attached models?

The critical examples are:
- Modelica.Mechanics.Rotational.Examples.GenerationOfFMUs
- ModelicaTest.MultiBody.PlanarLoopWithMove

Other Move examples are here:
- ModelicaTest.Rotational.TestMove
- ModelicaTest.Rotational.AllComponents
- ModelicaTest.Rotational.TestBearingConversion
